packages/errors/src/index.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * Creates a new error type with the given name and parent class. Sets up
+ * boilerplate necessary to successfully subclass Error and preserve stack trace
+ * @param name - The name of the error type
+ * @param ParentType - The error that serves as the parent
+ * @returns The new type
+ */
+export const createErrorType = (
+  name: string,
+  ParentType: new (message: string) => Error = Error
```

Comment:
Whoa. Now this I can't quite follow

Reply:
Just wait, things are going to get weirder the higher up the stack we go.

Reply:
Here are the docs for what I'm doing here: https://www.typescriptlang.org/v2/docs/handbook/generics.html

Basically, I'm saying that the `ParentType` argument should be a type that is a constructor function (a class) that takes a single argument that's a string and returns an instance of an Error. The default is set to Error.

Reply:
Incidentally, [Typescript Weekly](https://www.typescript-weekly.com/) just included an article about this: https://2ality.com/2020/04/classes-as-values-typescript.html